File: docs/lab1alt/code/MainEval.java
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;
import java.io.IOException;
import java.util.List;
import java.util.Iterator;

public class MainEval {
    public static void main(String[] args) throws IOException {
        CharStream  input   = CharStreams.fromStream(System.in);
        Lexer       lexer   = new ExprLexer(input);
        TokenStream tokens  = new CommonTokenStream(lexer);
        ExprParser  parser  = new ExprParser(tokens);
        ParseTree   tree    = parser.expr(); // 取得一颗以 expr 为根的 AST
        EvalVisitor visitor = new EvalVisitor();
        System.out.println(tree.accept(visitor));
    }
}


class EvalVisitor extends ExprBaseVisitor<Integer> {
    @Override public Integer visitAdd(ExprParser.AddContext ctx) {
        // ctx.add() 返回 add 子结点。
        if (ctx.add() == null)
            // 这个结点是 add: mul;
            return visitChildren(ctx);
        // 这个结点是 add: add op=(Add|Sub) mul;
        int lhs = ctx.add().accept(this);
        int rhs = ctx.mul().accept(this);
        // 访问 op 子结点不是 ctx.op() 而是 ctx.op.
        // 用 .getText() 取得字符串
        switch (ctx.op.getText()) {
            case "+": return lhs + rhs;
            case "-": return lhs - rhs;
        }
        return 0; // unreachable
    }

    @Override public Integer visitMul(ExprParser.MulContext ctx) {
        // 包含所有 atom 子结点
        List<ExprParser.AtomContext> atoms = ctx.atom();
        List<ExprParser.MulOpContext> ops = ctx.mulOp();
        int s = atoms.get(0).accept(this);
        Iterator<ExprParser.AtomContext> itAtom = atoms.iterator();
        Iterator<ExprParser.MulOpContext> itOps = ops.iterator();
        for (itAtom.next(); itAtom.hasNext();) {
            // 使用 .getText() 取得字符串
            if (itOps.next().getText().equals("*"))
                s *= itAtom.next().accept(this);
            else
                s /= itAtom.next().accept(this);
        }
        return s;
    }

    @Override public Integer visitAtomParen(ExprParser.AtomParenContext ctx) {
        // visitChilren 返回的是 ctx.Rparen().accept(this)
        return ctx.expr().accept(this);
    }

    @Override public Integer visitAtomInteger(ExprParser.AtomIntegerContext ctx) {
        return Integer.parseInt(ctx.Integer().getText());
    }
}

File: docs/lab1alt/code/maineval.py
import sys
from antlr4 import *
from ExprLexer import ExprLexer
from ExprParser import ExprParser
from ExprVisitor import ExprVisitor


class EvalVisitor(ExprVisitor):
    def visitAdd(self, ctx:ExprParser.AddContext):
        # ctx.add() 返回 add 子结点。
        if ctx.add() is None:
            # 这个结点是 add: mul;
            return self.visitChildren(ctx)
        # 这个结点是 add: add op=(Add|Sub) mul;
        lhs = ctx.add().accept(self)
        rhs = ctx.mul().accept(self)
        # 访问 op 子结点不是 ctx.op() 而是 ctx.op.
        # type(ctx.op) == Token，所以用 .text 取得字符串
        #（这是 Python API 一个不完善的地方, 本来应该用 .getText() 的）
        if ctx.op.text == '+':
            return lhs + rhs
        else:
            return lhs - rhs

    def visitMul(self, ctx:ExprParser.MulContext):
        # 返回一个 list，包含所有 atom 子结点
        atoms = ctx.atom()
        ops = ctx.mulOp()
        s = atoms[0].accept(self)
        for i in range(len(ops)):
            # ops[i] 是非终结符 mulOp，类型是 MulOpContext 继承(ParserRuleContext)
            # 所以用 .getText() 取得字符串
            if ops[i].getText() == '*':
                s *= atoms[i+1].accept(self)
            else:
                s //= atoms[i+1].accept(self)
        return s

    def visitAtomParen(self, ctx:ExprParser.AtomParenContext):
        # visitChilren 返回的是 ctx.Rparen().accept(self)
        return ctx.expr().accept(self)

    def visitAtomInteger(self, ctx:ExprParser.AtomIntegerContext):
        # ctx.Integer() 类型是 TerminalNodeImpl
        # 也用 .getText() 取得字符串
        return int(ctx.Integer().getText())
        # return int(ctx.getText()) # 和上面等价的


def main():
    input   = InputStream(sys.stdin.read())
    lexer   = ExprLexer(input)
    tokens  = CommonTokenStream(lexer)
    parser  = ExprParser(tokens)
    tree    = parser.expr() # 取得一颗以 expr 为根的 AST
    visitor = EvalVisitor()
    print(tree.accept(visitor))


if __name__ == '__main__':
    main()

